NF: `configuration` command

mih opened the pull request:
This closes a long-standing feature gap: a command that can
query and modify configuration across all scopes that are
relevant for DataLad, and also works recursively.

Importantly, it can render an annotated configuration dump,
as a convenient way to determine available configuration options
and effective settings.

try running: `datalad configuration`
![image](https://user-images.githubusercontent.com/136479/103644051-01bedf00-4f56-11eb-8d2d-b95f84a61448.png)

The choice of name is somewhat forced by the presence of `.config`
object in every Dataset instance.

Fixes gh-4359

yarikoptic commented:
Sorry, I am missing why not just `datalad config` ???

mih commented:
> Sorry, I am missing why not just `datalad config` ???

How would you implement the transition? I saw and see no meaningful approach. I have no further reason.

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+            constraints=EnsureChoice(*config_actions)),
+        scope=Parameter(
+            args=("--scope",),
+            doc="""identify to configuration scope for getting or setting
```
`s/indentify to/s/indentify/` or `s/identify to //` ?

mih replied:
Fixed!

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+            or name/value pair (for action 'set')""",
+            nargs='*',
+            metavar='name[=value]',
+            constraints=EnsureStr() | EnsureNone()),
```
For the python docs, this leads to a confusing `: <type>`:

```
    spec : sequence of str or None, optional
      configuration name (for actions 'get' and 'unset'), or name/value
      pair (for action 'set'). [Default: None]
```

So, in addition to adding a `[PY: ... <can use tuple> PY]` bit to this description, perhaps it'd be better to just drop the `constraints` argument, as `copy_file` does for `specs_from`.

mih replied:
Dropped. Thx

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+    )
+
+    @staticmethod
+    @datasetmethod(name='cfg')
```
I'm guessing this is a stale part from a previous iteration and meant to match the unbound name, `configuration`.

mih replied:
Fixed. Thx!

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+                    ui.message('\n'.join(wrap(
+                        s,
+                        initial_indent='# ',
+                        subsequent_indent='# ',
```
I think dumping this annotation very helpful, and I like the format.  I wonder if we should move the title/text extraction to `_dump`, letting the custom rerender just deal with the formatting.  That way the other renderers' output (I'm particularly interested in `--json{,-pp}`) would be more helpful.

mih replied:
Yes, great idea!

mih replied:
Done

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+        if specs:
+            lgr.warning('Provided configuration selection is ignored when dumping')
+        # dumping is querying for all known keys
+        # TODO warn that specs will be ignored, if there are any
```
It looks like this todo is covered by the warning above.  However, I think the warning will be very easy to miss with the verbose output of `dump`.  I'd vote for changing this warning to a value error early in `__call__`.

Also, should a similar warning/error also happen with `--scope`?

mih replied:
Yes, you are right. Done.

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+                status='error',
+                message=(
+                    "Configuration key without a section: '%s'",
+                    spec[0],
```
Processing this result fails:

```
[ERROR  ] unhashable type: 'dict' [utils.py:_process_results:565] (TypeError) 
```

This is because `res_kwargs` is incorrectly stored as the `action`.  One fix would be using `**res_kwargs` and dropping the explicit `logger` argument.

mih replied:
Oh lord. Thx!

mih replied:
Done.

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+        #elif action == 'get-all':
+        #    res = _get_all(cfg, scope, spec)
+        elif action == 'set':
+            res = _set(cfg, scope, *spec)
```
If the user does a `git config`-inspired `set a.b.c val`, this will
fail with

```
[ERROR  ] _set() missing 1 required positional argument: 'value' [configuration.py:configuration:275] (TypeError)
```

It seems worth giving a more informative error message since I think mixing the two forms will be a common mistake.

mih replied:
```
% datalad configuration set mike.tyson winner
[ERROR  ] Values must be provided for all configuration settings. Missing: ['mike.tyson', 'winner'] [configuration.py:__call__:179] (ValueError) 
```

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+        # if we have no value, but a key (i.e. in dump mode)
+        cfg_defs.get(name, {}).get('default', None))
+    return dict(
+        action='get_configuration',
```
Should the action be 'dump_configuration' for consistency?

mih replied:
Yes, done.

kyleam commented on datalad/core/local/configuration.py:
```diff
@@ -0,0 +1,337 @@
+        if scope else cfg.get(
+            name,
+            # pull a default from the config definitions
+            # if we have no value, but a key (i.e. in dump mode)
```
Is the "(i.e. in dump mode)" a copy-paste error?

mih replied:
Yes, a left-over. Fixed.

kyleam reviewed (commented):
Thanks, looks nice/useful.

mih commented:
I think this is ready. Notable changes since @kyleam looked at it:

- move from `core.local` to `local`, and added a note about its young age and potential changes
- basic tests that at least run all lines
- added to docs
- more comprehensive argument evaluation
- minimized functionality implemented in the result renderer and moved config description reporting to the result values

kyleam commented on setup.py:
```diff
@@ -28,6 +28,9 @@
         'colorama; platform_system=="Windows"',
+        # for as long as we support python 3.5
+        # to prevent "UnicodeEncodeError: 'charmap' codec can't encode..."
+        'win-unicode-console; platform_system=="Windows"',
```
`python_requires='>=3.6'` is already declared below.

mih replied:
Yeah, this lead to nothing. I removed it in full.

kyleam commented:
>  Notable changes [...]

Thanks, the summary is very appreciated.

The updates look good to me.  I don't have anything useful to say about the windows encoding error, though I think the unsuccessful attempt in "BF: Enable unicode input output on windows with Python < 3.6" should be discarded given we already require Python 3.6.

adswa commented:
(FTR: feel free to change the .git/config user config on the shared Windows computer, @mih. You are committing with my identity :P)

mih commented:
> (FTR: feel free to change the .git/config user config on the shared Windows computer, @mih. You are committing with my identity :P)

Ha, and I always thought "I like her style" ;-)

mih commented:
OK, last push. I fixed up the commit ownership @adswa, brought coverage up to 100% again.
The travis errors are all #5288.

```
(datalad3-dev) mih@meiner ~/hacking/datalad/git (git)-[bf-4359] % python -m nose -s -v datalad.local.tests.test_configuration:test_something --with-coverage --cover-package datalad.local.configuration --cover-erase
datalad.local.tests.test_configuration.test_something ... 4
ok
Versions: annexremote=1.4.3 appdirs=1.4.4 boto=2.49.0 cmd:7z=16.02 cmd:annex=8.20201103 cmd:bundled-git=UNKNOWN cmd:git=2.29.2 cmd:system-git=2.29.2 cmd:system-ssh=8.1p1 humanize=2.6.0 iso8601=0.1.13 keyring=21.4.0 keyrings.alt=3.4.0 msgpack=1.0.0 requests=2.24.0 scrapy=2.4.1 wrapt=1.12.1
Obscure filename: str=b' |;&%b5{}\'"\xce\x94\xd0\x99\xd7\xa7\xd9\x85\xe0\xb9\x97\xe3\x81\x82 .datc ' repr=' |;&%b5{}\'"ΔЙקم๗あ .datc '
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='UTF-8'
Environment: LANG='en_US.UTF-8' LANGUAGE='en_US:en' PATH='/home/mih/env/datalad3-dev/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/home/mih/bin:/home/mih/bin/pycharm/bin' GIT_EDITOR='vim' LC_CTYPE='en_US.UTF-8' GIT_CONFIG_PARAMETERS="'init.defaultBranch=dl-test-branch'"

Name                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------
datalad/local/configuration.py     129      0   100%
----------------------------------------------------------------------
Ran 1 test in 2.375s

OK
```

mih commented:
Verdict from developer call: merge and see.
